graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename types, interfaces, enums, etc
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+        scalars: set of strings, the set of user defined scalars
+
+    Returns:
+        Dict[str, str], the new type name to original type name map
+
+    Raises:
+        SchemaError if the rename causes name conflicts
+    """
+    visitor = RenameSchemaVisitor(rename_func, query_type, scalars)
+    visit(ast, visitor)
+
+    return visitor.reverse_name_map
+
+
+def _rename_root_fields(ast, rename_func, query_type):
+    """Rename root fields, aka fields of the query type.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename fields of the query type
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+
+    Returns:
+        Dict[str, str], the new root field name to original root field name map
+
+    Raises:
+        SchemaError if rename causes root field name conflicts
+    """
+    visitor = RenameRootFieldsVisitor(rename_func, query_type)
+    visit(ast, visitor)
+
+    return visitor.reverse_field_map
+
+
+class RenameSchemaVisitor(Visitor):
+    """Traverse a Document AST, editing the names of nodes."""
+    def __init__(self, rename_func, query_type, scalar_types):
+        self.rename_func = rename_func  # callable that takes string to string
+        self.reverse_name_map = {}  # Dict[str, str], from new name to original name
+        self.query_type = query_type  # str
+        self.scalar_types = frozenset(scalar_types)
+        self.builtin_types = frozenset({'String', 'Int', 'Float', 'Boolean', 'ID'})
+
+        # NOTE: behavior of Directive type is not completely clear
+        self.noop_types = frozenset({
+            'Name', 'Document', 'Argument', 'IntValue', 'FloatValue', 'StringValue',
+            'BooleanValue', 'EnumValue', 'ListValue', 'Directive', 'ListType', 'NonNullType',
+            'SchemaDefinition', 'OperationTypeDefinition', 'ScalarTypeDefinition',
+            'FieldDefinition', 'InputValueDefinition', 'EnumValueDefinition',
+            'DirectiveDefinition'
+        })
+        self.rename_types = frozenset({
+            'NamedType', 'ObjectTypeDefinition', 'InterfaceTypeDefinition',
+            'UnionTypeDefinition', 'EnumTypeDefinition'
+        })
+        self.unexpected_types = frozenset({
+            'OperationDefinition', 'SelectionSet', 'Field', 'FragmentSpread',
+            'InlineFragment', 'FragmentDefinition'
+        })
+
+    def _rename_name_add_to_record(self, node):
+        """Rename the value of the node, and add the name mapping to reverse_name_map.
+
+        Don't rename if the type is the query type, a scalar type, or a builtin type.
+
+        Args:
+            node: Name type Node
```

Comment:
`Name type Node`?

Reply:
The input node is of class `Name` that inherits from class `Node`, maybe I should just say that the type is Name?

Reply:
With such a generic name, a somewhat longer explanation might be useful. The objective here is to be informative and useful even to an unfamiliar, casual reader of the code. To such a reader, `Name` type is not very informative, and `inherits from class Node` doesn't provide meaningful new information either because they are probably not familiar with the `Node` class.

In such a situation, explain what the significance of the object is, and how it is being operated on in this function. For example: `Name AST object, describing the name of an AST element such as a type, interface, or scalar.` plus some description of what will happen to it in this function.